rinja_derive/src/config.rs

```diff
@@ -293,27 +297,36 @@ fn str_set<'a>(vals: &[&'a str]) -> Vec<Cow<'a, str>> {
 }
 
 pub(crate) fn get_template_source(
-    tpl_path: &Path,
+    tpl_path: &Rc<Path>,
     import_from: Option<(&Rc<Path>, &str, &str)>,
 ) -> Result<Rc<str>, CompileError> {
-    match fs::read_to_string(tpl_path) {
-        Ok(mut source) => {
-            if source.ends_with('\n') {
-                let _ = source.pop();
+    thread_local! {
+        static CACHE: RefCell<HashMap<Rc<Path>, Rc<str>>> = RefCell::new(HashMap::new());
+    }
+    CACHE.with(|cache| {
+        let mut cache = cache.borrow_mut();
+        let entry = match cache.entry(tpl_path.clone()) {
+            Entry::Vacant(entry) => entry,
+            Entry::Occupied(source) => return Ok(source.get().clone()),
+        };
+        let mut source = match fs::read_to_string(tpl_path) {
+            Ok(source) => source,
+            Err(err) => {
+                let msg = format!(
+                    "unable to open template file '{}': {err}",
+                    tpl_path.to_str().unwrap(),
+                );
+                let file_info = import_from.map(|(node_file, file_source, node_source)| {
+                    FileInfo::new(node_file, Some(file_source), Some(node_source))
+                });
+                return Err(CompileError::new(msg, file_info));
             }
-            Ok(source.into())
-        }
-        Err(err) => {
-            let msg = format!(
-                "unable to open template file '{}': {err}",
-                tpl_path.to_str().unwrap(),
-            );
-            let file_info = import_from.map(|(node_file, file_source, node_source)| {
-                FileInfo::new(node_file, Some(file_source), Some(node_source))
-            });
-            Err(CompileError::new(msg, file_info))
+        };
+        if source.ends_with('\n') {
+            let _ = source.pop();
         }
-    }
+        Ok(entry.insert(source.into()).clone())
```

Comment:
It'd be better to replace `clone` calls (here and above) with `Rc::clone`. Not useful for performance, but better for reading. :)